src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicAnalysisDataReader.cs

```diff
@@ -0,0 +1,367 @@
+            {
+                throw new BadImageFormatException();
+            }
+
+            _guidHeapBlob = new Blob(buffer + guidHeapOffset, guidHeapSize);
+            _blobHeapBlob = new Blob(buffer + blobHeapOffset, blobHeapSize);
+        }
+
+        public static DynamicAnalysisDataReader TryCreateFromPE(PEReader peReader)
+        {
+            // TODO: review all range checks, better error messages
+
+            var mdReader = peReader.GetMetadataReader();
+            long offset = -1;
+            foreach (var resourceHandle in mdReader.ManifestResources)
+            {
+                var resource = mdReader.GetManifestResource(resourceHandle);
+                if (resource.Implementation.IsNil &&
+                    resource.Attributes == ManifestResourceAttributes.Private &&
+                    mdReader.StringComparer.Equals(resource.Name, "<DynamicAnalysisData>"))
+                {
+                    offset = resource.Offset;
+                }
+            }
+
+            if (offset < 0)
+            {
+                return null;
+            }
+
+            var resourcesDir = peReader.PEHeaders.CorHeader.ResourcesDirectory;
+            if (resourcesDir.Size < 0)
+            {
+                throw new BadImageFormatException();
+            }
+
+            int start;
+            if (!peReader.PEHeaders.TryGetDirectoryOffset(resourcesDir, out start))
+            {
+                return null;
+            }
+
+            var peImage = peReader.GetEntireImage();
+            if (start >= peImage.Length - resourcesDir.Size)
+            {
+                throw new BadImageFormatException();
+            }
+
+            byte* resourceStart = peImage.Pointer + start;
+            int resourceSize = *(int*)resourceStart;
+            if (resourceSize > resourcesDir.Size - sizeof(int))
+            {
+                throw new BadImageFormatException();
+            }
+
+            return new DynamicAnalysisDataReader(resourceStart + sizeof(int), resourceSize);
+        }
+
+        private static int ReadReference(ref BlobReader reader, bool smallRefSize)
+        {
+            return smallRefSize ? reader.ReadUInt16() : reader.ReadInt32();
+        }
+
+        public ImmutableArray<LinePositionSpan> GetSpans(BlobHandle handle)
+        {
+            var builder = ArrayBuilder<LinePositionSpan>.GetInstance();
+
+            var reader = GetBlobReader(handle);
+
+            // header:
+            int documentRowId = ReadDocumentRowId(ref reader);
+            int previousStartLine = -1;
+            ushort previousStartColumn = 0;
+
+            // records:
+            while (reader.RemainingBytes > 0)
+            {
+                int deltaLines, deltaColumns;
+                ReadDeltaLinesAndColumns(ref reader, out deltaLines, out deltaColumns);
+
+                // document:
+                if (deltaLines == 0 && deltaColumns == 0)
```

Comment:
Yes. line == 0 and column == 0 is an "escape" sequence for document change. Zero-width spans are not allowed.